src/cpp/benders/merge_mps/MergeMPS.cpp

```diff
@@ -1,269 +1,368 @@
-    if (optimality_l)
-    {
-        sol_infos.problem_status = "OPTIMAL";
-    }
-    else
+    sol_infos.solution.candidates.clear();
+    sol_infos.solution.candidates.reserve(investments.size());
+
+    for (const auto& [var_name, var_value]: investments)
     {
-        sol_infos.problem_status = "ERROR";
+        const Output::CandidateData candidate_data{var_name, var_value, -1, -1};
+        sol_infos.solution.candidates.push_back(candidate_data);
     }
 
+    sol_infos.problem_status = is_sol_optimal ? "OPTIMAL" : "ERROR";
+
     _writer->update_solution(sol_infos);
     _writer->dump();
 }
 
+MergeMasterSubproblemMPS::MergeMasterSubproblemMPS(MergeMPSOptions options,
+                                                   Logger logger,
+                                                   std::shared_ptr<Output::OutputWriter> writer):
+    AbstractMergeMPS(options, logger, writer)
+{
+}
+
 /*!
- *  \brief Return slave weight value
+ *  \brief Return subproblem weight value
  *
- *  \param nslaves : total number of slaves
+ *  \param nb_subproblems : total number of subproblems
  *
- *  \param name : slave name
+ *  \param name : subproblem name
  */
-double MergeMPS::slave_weight(int nslaves, const std::string& name) const
+double MergeMasterSubproblemMPS::get_objective_weight(const int nb_subproblems,
+                                                      const std::string& name) const
 {
     if (_options.SLAVE_WEIGHT == SUBPROBLEM_WEIGHT_UNIFORM_CST_STR)
     {
-        return 1 / static_cast<double>(nslaves);
+        return 1.0 / nb_subproblems;
     }
     else if (_options.SLAVE_WEIGHT == SUBPROBLEM_WEIGHT_CST_STR)
     {
-        const double weight(_options.SLAVE_WEIGHT_VALUE);
-        return 1 / weight;
+        return 1.0 / _options.SLAVE_WEIGHT_VALUE;
     }
     else
     {
-        if (_options.weights.find(name) == _options.weights.end())
+        const auto found = _options.weights.find(name);
+        if (found == _options.weights.end())
         {
             _logger->display_message("No weight found for " + name
                                        + ". Problem will not contribute to objective function",
                                      LogUtils::LOGLEVEL::WARNING,
                                      "MergeMPS");
+            return 0.;
+        }
+        return found->second;
+    }
+}
+
+/**
+ * \brief Add coupling equality constraints between subproblems
+ *
+ * \param candidates : Mapping of investment candidates
+ */
+void MergeMasterSubproblemMPS::add_coupling_constraints(const CouplingMap& candidates)
+{
+    // TODO Investigate why following check
+    // TODO creates a segfault when structure.txt is empty
+    // if (candidates.empty())
+    // {
+    //     return;
+    // }
+
+    size_t nb_elem_reserve{0}; // 2-permutation of variables
+    for (const auto& [_, file_mapping]: candidates)
+    {
+        nb_elem_reserve += file_mapping.size() * (file_mapping.size() - 1);
```

Comment:
For each variable, if we have n subproblems in the `file_mapping`, we add n-1 coupling constraints. Each of the constraints has 2 elements. Therefore, I would have only allocated `2 * (file_mappping.size() - 1)` am I wrong ?
We could simply run in debug mode and see what is the value of `nelem` at the end of this function and compare with `nb_eleme_reserve`

Reply:
I agree with you, but this is the current behavior and I didn't want to change it.
But we can test it and change it if your suspicions are right